### DISCUSSION TITLE
x-model updates unexpected data

### DISCUSSION THREAD

**Discussion by andrewheberle** [alpinejs/alpine#3321] [1. Help] [answered]
I am using "x-model" in order to populate form controls in a modal dialog to allow editing of existing values, with a click on save or cancel causing an API call or closing the modal respectively.

What I'm seeing though is that changes to the input field is immediately reflected in the underlying data despite having passed this as an argument to another "x-data" entry.

Here is a limited example:

```js
Alpine.data('list', () => ({
    init() {
        // set up data (would come from api normally)
        this.data = [
            { name: 'bob' },
            { name: 'fred' },
            { name: 'bill' },
        ];
    },
    data: Alpine.$persist([]),
}));

Alpine.data('item', ({index, item}) => ({
    data: Alpine.$persist([]),
    item: item,
    open: false,
    save() {
        // would update api normally, but update in-place for test
        this.data[index] = this.item;
        this.open = false;
    },
    close() {
        this.open = false;
    },
}));
```

With the HTML:

```html
<div x-data="list">
    <table>
        <template x-for="(entry, index) in data">
            <tr x-data="item({id: index, item: entry})">
                <td x-text="item.name"></td>
                <td @click="open = true">Edit</td>
                <template x-teleport="body">
                    <div x-id="['input']" x-show="open" @click.outside="close">
                        <label for="$id('input')">Name:</label>   
                        <input type="text" :id="$id('input')" x-model="item.name">
                        <button @click="save">Save</button>
                        <button @click="close">Cancel</button>
                    </div>
                </template>
            </tr>
        </template>
    </table>
</div>
```

The issue I'm trying to solve is a simple modal that will allow editing my list of items, but only to have changes made if the user hits "save", rather than updating the current value and where it came from.

**Comment by ekwoka** [comment, marked as answer]
It's javascript. Objects are not cloned as they're passed around, they are passed by reference.

So you're passing the same literal object around and updating it. You'd need to clone the object if you want it to only update the state within itself and then pass the data up only when you choose to.

**Reply by andrewheberle** [reply]
I had a moment where I thought all my problems were solved after reading your message...but I may have come up against the limitations of my Javascript knowledge.

I have updated part of my test case so it looks like this:

```js
Alpine.data('item', ({index, item}) => ({
    init() {
        let clone = Object.assign({}, item);
        this.item = clone;
    },
    data: Alpine.$persist([]),
    item: {},
    open: false,
    save() {
        this.data[index] = this.item;
        this.open = false;
    },
    close() {
        this.open = false;
    },
}));
```

Which based on my understanding should mean the "item" object is being cloned by using `Object.assign`...however this still seems to not be the case.

**Reply by SimoTod** [reply]
It is cloned but you have several small typos in your first example.

`<td x-text="item.name"></td>` <-- you are printing out the variable of the subcomponent rather than the main one, it should be `<td x-text="entry.name"></td>`

`Alpine.data('item', ({index, item}) => ({` your named parameter should be `id` not `index`

I'm not sure why you need `data: Alpine.$persist([]),` in the subcomponent, it has already access to data from the parent

You technically don't need to use init in the subcomponent

https://codepen.io/SimoTod/pen/rNKPqRj

**Reply by andrewheberle** [reply]
Oh gosh...of course...thanks for pointing out that silly mistake!
